AmplifyPlugins/API/Sources/AWSAPIPlugin/AppSyncRealTimeClient/AppSyncRealTimeClient.swift

```diff
@@ -0,0 +1,457 @@
+            log.debug("[AppSyncRealTimeClient] client is already connecting or connected")
+            return
+        case .disconnecting:
+            try await waitForState(.disconnected)
+        case .connectionDropped, .disconnected, .none:
+            break
+        }
+
+        guard self.state.value != .connecting else {
+            log.debug("[AppSyncRealTimeClient] actor reentry, state has been changed to connecting")
+            return
+        }
+
+        self.state.send(.connecting)
+        log.debug("[AppSyncRealTimeClient] client start connecting")
+
+        try await RetryWithJitter.execute { [weak self] in
+            guard let self else { return }
+            try await AppSyncRealTimeRequest.sendRequest(
+                request: .connectionInit,
+                responseStream: subject.eraseToAnyPublisher()
+            ) { [weak self] _ in
+                await self?.webSocketClient.connect(
+                   autoConnectOnNetworkStatusChange: true,
+                   autoRetryOnConnectionFailure: true
+                )
+            }
+        }
+    }
+
+    /**
+     Disconnect only when there are no subscriptions exist.
+     */
+    func disconnectWhenIdel() async {
+        if self.subscriptions.isEmpty {
+            log.debug("[AppSyncRealTimeClient] no subscription exist, client is trying to disconnect")
+            await disconnect()
+        } else {
+            log.debug("[AppSyncRealTimeClient] client only try to disconnect when no subscriptions exist")
+        }
+    }
+
+    /**
+     Disconnect from AppSync real-time server.
+     */
+    func disconnect() async {
+        guard self.state.value != .disconnecting else {
+            log.debug("[AppSyncRealTimeClient] client already disconnecting")
+            return
+        }
+
+        defer { self.state.send(.disconnected) }
+
+        log.debug("[AppSyncRealTimeClient] client start disconnecting")
+        self.state.send(.disconnecting)
+        self.cancellablesBindToConnection = Set()
+        await self.webSocketClient.disconnect()
+        log.debug("[AppSyncRealTimeClient] client is disconnected")
+    }
+
+    /**
+     Subscribing to a query with unique identifier.
+     - Parameters:
+        - id: unique identifier
+        - query: GraphQL query for subscription
+
+     -  Returns:
+        A never fail data stream for AppSyncSubscriptionEvent.
+     */
+    func subscribe(id: String, query: String) async throws -> AnyPublisher<AppSyncSubscriptionEvent, Never> {
+        log.debug("[AppSyncRealTimeClient] Received subscription request id: \(id), query: \(query)")
+        let subscription = AppSyncRealTimeSubscription(id: id, query: query, endpoint: endpoint)
+        subscriptions[id] = subscription
+
+
+        // Placing the actual subscription work in a deferred task and
+        // promptly returning the filtered publisher for downstream consumption of all error messages.
+        defer {
```

Comment:
I'm trying to understand the point of `defer` here, does it actually do anything? for example is`defer` block executed before or after this line?
```
return filterAppSyncSubscriptionEvent(with: id)
            .merge(with: (await subscription.publisher).toAppSyncSubscriptionEventStream())
            .eraseToAnyPublisher()
```

If executed after the line, then the code execution would look like this:
```
let publisher = filterAppSyncSubscriptionEvent(with: id)
            .merge(with: (await subscription.publisher).toAppSyncSubscriptionEventStream())
            .eraseToAnyPublisher()

// run defer block

return publisher
```

if executed before, then the `defer` usage seems unnecessary

Reply:
The order is important here for downstream to receive all the related events. `defer` here is to ensure the block of code start running after this function returns. 

The suggested code snippet should be working as well. As the `subscribe` API is marked as async as well, I just suspect there is a small chance that the defer Task could be ran simultaneously with `suscribe`.